src/snowflake/cli/plugins/nativeapp/codegen/compiler.py

```diff
@@ -0,0 +1,73 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Dict, Optional
+
+from snowflake.cli.api.project.schemas.native_app.native_app import NativeApp
+from snowflake.cli.api.project.schemas.native_app.path_mapping import (
+    PathMapping,
+    ProcessorMapping,
+)
+from snowflake.cli.plugins.nativeapp.codegen.artifact_processor import (
+    ArtifactProcessor,
+    UnsupportedArtifactProcessorError,
+)
+from snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor import (
+    SnowparkAnnotationProcessor,
+)
+
+SNOWPARK_PROCESSOR = "snowpark"
+
+# Singleton dictionary of all processors created and shared between different artifact objects
+cached_processors: Dict[str, ArtifactProcessor] = {}
```

Comment:
This is a code smell to me. Should we introduce some sort of context object or attach it to an existing one?

Reply:
I'd like to understand more on why this is a concerning issue.
Regardless, I could make this a part of the existing global cli context. However, most of the properties on the existing cli context class are universal, i.e. used across various command subgroups. But this processors dict is a very specific part of native apps and within native apps, the artifacts that it processes. So is adding it to a global context object justified?

If we want to leverage this new codegen implementation for wider use beyond NA going forward, e.g. in pure Snowpark, then it might make sense to promote it to be universal.

Reply:
We can certainly make a team decision that the current approach needs to be re-written if we want wider use. Generally I expect there to be one context-level object for each global-ish lifetime and for us to jam things in there. This isn't a problem right now because we're only ever working with one project.

My suggestion would be to call out the cases in which this cached mapping would cause incorrect behaviour in the comment above, I think we can accept it for now but we need to be mindful if/when we move out into a library.

Reply:
I've added more clarification, lmk.